docs: Add DK-Bench docs (#3124)
instructlab/instructlab: docs/user/dk_bench.md

# DK-Bench (Domain-Knowledge Bench)

## Background

Current knowledge evaluation in InstructLab, MMLU-Branch, evaluates models on their answers to multiple choice questions.
In MMLU-Branch, the model either gets the answer right or wrong, but there's no in between to give the model credit on moderately correct or incorrect answers.

Additionally, there was no way to bring a custom evaluation data set to run on InstructLab trained models.

DK-Bench (Domain-Knowledge Bench) provides the ability to bring custom evaluation questions and score the models answers on a sliding scale.

## Description

> DK-Bench provides a flexible way for users to bring custom evaluation questions, reference answers, and collect a models answer to the questions.

Reviewer: ```suggestion
DK-Bench provides a flexible way for users to bring custom evaluation questions, reference answers, and collect a model's answers to the questions.
```

> The responses to each of questions is given a score from 1-5 by a judge model.

Reviewer: ```suggestion
The response to each question is given a score from 1-5 by a judge model.
```

The current default judge model is GPT-4o by OpenAI.

### User Input

> To run DK-Bench, users must provide a `.jsonl` file with every question they wish to ask a model to answer and evaluate.

Reviewer: ```suggestion
To run DK-Bench, users must provide a `.jsonl` file containing every question they wish to ask a model to answer and evaluate.
```

| Score | Criteria |
| ---------------- | -------- |
| 1 | The response is entirely incorrect, irrelevant, or does not align with the reference in any meaningful way |
| 2 | The response partially matches the reference but contains major errors, significant omissions, or irrelevant information |
| 3 | The response aligns with the reference overall but lacks sufficient detail, clarity, or contains minor inaccuracies |
| 4 | The response is mostly accurate, aligns closely with the reference, and contains only minor issues or omissions |
| 5 | The response is fully accurate, completely aligns with the reference, and is clear, thorough, and detailed |

### Output

After running DK-Bench, users will see a report with the question by question score, the average, and the total for their run.

```shell
# DK-BENCH REPORT

## MODEL: granite-7b-lab

Question #1:     5/5
Question #2:     5/5
Question #3:     5/5
Question #4:     5/5
Question #5:     2/5
Question #6:     3/5
Question #7:     2/5
Question #8:     3/5
Question #9:     5/5
Question #10:     5/5
----------------------------
Average Score:   4.00/5
Total Score:     40/50

> Responses and scores written to:

Reviewer: ```suggestion
Responses and scores are written to:
```

~/.local/share/instructlab/internal/eval_data/dk_bench/granite-7b-lab/results_2025-02-04T16:18:02.337010.jsonl
```

> The original question, the responses a model gave for that specific question, the reference answer, the score of the judge model are all output to an output file.

Reviewer: ```suggestion
The original question, the responses a model gave for that specific question, the reference answer, and the score of the judge model are all output to an output file.
```

> The output file could be `.jsonl` (the default and seen below), `.csv`, and `.xlsx`. This can be set using the `--output-file-formats` flag or in the `dk_bench` section of the configuration file.

Reviewer: ```suggestion
The output file can be `.jsonl` (the default and seen below), `.csv`, or `.xlsx`. This can be set using the `--output-file-formats` flag or in the `dk_bench` section of the configuration file.
```

```json
{"user_input":"What is the capital of Canada?","response":"The capital of Canada is Ottawa.","reference":"The capital of Canada is Ottawa.","scores":5}
{"user_input":"What is the capital of Mexico?","response":"The capital of Mexico is Mexico City.","reference":"The capital of Mexico is Mexico City.","scores":5}
```

## Examples

> [!NOTE]
> The DK-Bench evaluation requires the environment variables `OPENAI_API_KEY` to be set. It will not run if this variable is not set.
> To set your OpenAI API Key in the shell run:
> `export OPENAI_API_KEY=<API KEY HERE>`

> DK-Bench has many flags that are applied when the `ilab model evaluate` command to run DK-Bench.

Reviewer: ```suggestion
DK-Bench has many flags that are applied when the `ilab model evaluate` command is used to run DK-Bench.
```

I think? Is this sentence needed though?

# DK-Bench (Domain-Knowledge Bench)

> 

Reviewer 2: Seems it would be better to add a short introduction in `README.md` and link to user doc.

The output file could be `.jsonl` (the default and seen below), `.csv`, and `.xlsx`. This can be set using the `--output-file-formats` flag or in the `dk_bench` section of the configuration file.

```json
{"user_input":"What is the capital of Canada?","response":"The capital of Canada is Ottawa.","reference":"The capital of Canada is Ottawa.","scores":5}
{"user_input":"What is the capital of Mexico?","response":"The capital of Mexico is Mexico City.","reference":"The capital of Mexico is Mexico City.","scores":5}
```

## Examples

> [!NOTE]
> The DK-Bench evaluation requires the environment variables `OPENAI_API_KEY` to be set. It will not run if this variable is not set.
> To set your OpenAI API Key in the shell run:
> `export OPENAI_API_KEY=<API KEY HERE>`

DK-Bench has many flags that are applied when the `ilab model evaluate` command to run DK-Bench.

Here is the most basic example of using DK-Bench, by passing in `dk_bench` to the `--benchmark` flag of `ilab model evaluate`, `--input-questions` with the path to the .jsonl dataset, and the path to a local model via `--model`.

```shell

> ilab model evaluate --benchmark dk_bench --input-questions questions.jsonl --model ~/.cache/instructlab/models/instructlab/granite-7b-lab

Reviewer 3: silly question, 
can the `--input-questions` be a path to the .jsonl as well?

Author: Yes! This could have also been `--input-questions /home/user/path/to/questions.jsonl`

Reviewer 3: was just curious, ty!

INFO 2025-02-04 16:15:55,707 numexpr.utils:162: NumExpr defaulting to 16 threads.
...
Responses and scores written to:
~/dk-bench-results-dir/granite-7b-lab/results_2025-02-04T16:18:02.337010.csv
~/dk-bench-results-dir/granite-7b-lab/results_2025-02-04T16:18:02.337010.xlsx

INFO 2025-02-04 16:18:02,337 instructlab.model.evaluate:176: ᕦ(òᴗóˇ)ᕤ Model Evaluation with DK-Bench completed! ᕦ(òᴗóˇ)ᕤ
```

## Configurations & Flags for Advanced Users

There are also flags for temperature, system prompt, and a changing the judge model for users interested in seeing the difference those settings make to evaluation results.

The `--temperature` flag sets the temperature for the model when it gets response for the questions. The default is `0.0`.
The `--system-prompt` flag sets the system prompt for the model when it gets response for the questions. The default system prompt depends on the model architecture of the model being evaluated.
The `--judge-model` flag sets the system prompt for the model when it gets response for the questions. The default is `gpt-4o`. The judge model needs to be an Open AI model.

> All of the flags for DK-Bench have variables that can be set in the config file as well.

Reviewer 4: ```suggestion
All of the flags for DK-Bench can be set in the config file as well.
```

> ## Running DK-Bench with generated responses

Reviewer 4: ```suggestion
## Running DK-Bench from pre-generated responses
```

> DK-Bench can be used with responses already provided in the `.jsonl` input file. If responses are already provided then no model needs to be provided as an argument to `ilab model evaluate`.

Reviewer 4: Wording

```suggestion
If you've already generated the answers ahead of time, or would like to avoid spinning up a student model for this benchmark, you can also pass a `.jsonl` file containing the model's responses in the `response` field.
```

```json
{"user_input":"What is the capital of Canada?","response":"The capital of Canada is Toronto.","reference":"The capital of Canada is Ottawa."}
{"user_input":"What is the capital of Mexico?","response":"The capital of Mexico is Mexico City.","reference":"The capital of Mexico is Mexico City."}

> ```

Reviewer 4: We should move this section below the code section below so readers first see the command they'd run, and then what structure their input file should be.
